File: Sources/Engine/editor/SKKEntryRemoveEditor.swift
//
//  SKKEntryRemoveEditor.swift
//  AquaSKKEngine
//
//  Created by mzp on 2025/03/05.
//

import AquaSKKBackend

public class SKKEntryRemoveEditorImpl: SKKEditorProtocol {
    let context: SKKInputContext
    var input: String
    var prompt: String
    var entry: SKKEntryBridge
    var candidate: SKKCandidateBridge

    public init(context: SKKInputContext) {
        self.context = context
        input = ""
        prompt = ""
        entry = SKKEntryBridge()
        candidate = .init()
    }

    public func readContext() {
        entry = context.entry
        candidate = context.candidate
        input.removeAll()
        prompt = "\(entry.entryString) /\(candidate.stringValue)/ を削除しますか？(yes/no) "
    }

    public func writeContext() {
        context.output.clear()
        context.output.compose(string: prompt + input)
        context.entry = entry
    }

    public func input(ascii: String) {
        input += ascii
    }

    public func input(fixed: String, input _: String, code _: Int) {
        input += fixed
    }

    public func bridgeInputEvent(_ rawValue: UInt32) {
        let event = SKKBaseEditorEvent(rawValue: rawValue)!
        inputEvent(event: event)
    }

    public func inputEvent(event: SKKBaseEditorEvent) {
        if event == .SKKBaseEditorEventBackSpace, !input.isEmpty {
            input.removeLast()
        }
    }

    public func commit(queue _: String) -> String {
        if input == "yes" {
            context.needs_setback = true
        } else {
            SKKBackendImpl.shared().remove(entry: entry.copy(), candidate: candidate.copy())
        }
        return ""
    }
}

File: Sources/Engine/editor/SKKEntryRemoveEditor.cpp
/* -*- C++ -*-

  MacOS X implementation of the SKK input method.

  Copyright (C) 2008 Tomotaka SUWA <t.suwa@mac.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#import <AquaSKKEngine/AquaSKKEngine-Preamble.h>
#import <AquaSKKEngine/SKKEntryRemoveEditor.h>
#import <AquaSKKEngine/SKKInputContext.h>
#import <AquaSKKEngine/AquaSKKEngine-Swift.h>

SKKEntryRemoveEditor::SKKEntryRemoveEditor(SKKInputContext *context)
    : SKKBaseEditor(context), impl_(new SwiftObject(AquaSKKEngine::SKKEntryRemoveEditorImpl::init(context))) {}

void SKKEntryRemoveEditor::ReadContext() {
    (*impl_)->readContext();
}

void SKKEntryRemoveEditor::WriteContext() {
    (*impl_)->writeContext();
}

void SKKEntryRemoveEditor::Input(const std::string &ascii) {
    (*impl_)->input(ascii);
}

void SKKEntryRemoveEditor::Input(const std::string &fixed, const std::string &input, char code) {
    (*impl_)->input(fixed, input, code);
}

void SKKEntryRemoveEditor::Input(SKKBaseEditorEvent event) {
    (*impl_)->inputEvent(event);
}

void SKKEntryRemoveEditor::Commit(std::string &queue) {
    queue = (*impl_)->commit(queue);
}
